### ISSUE TITLE
"Complete" feature gets added automatically & after installation of MSI, an empty feature created even if it is not selected

### ISSUE THREAD

**Issue by abskulkarni** [oleg-shilo/wixsharp#1578]
Hi @oleg-shilo 
I am attaching my Wix Project (InsuranceDataRepository.zip) & source directory for it (PROPHET_IDR_Build.zip) here.

I am facing two problems:

1. One feature named "Complete" gets added in Feature list of Feature free when MSI is run.
2. If I install MSI, it installs empty "IDRDatabaseScripts" folder (which is a feature) even if it is not selected.

Can you please guide me is in finding root cause & resolving these problems?

[PROPHET_IDR_Build.zip](https://github.com/user-attachments/files/16118325/PROPHET_IDR_Build.zip)
[InsuranceDataRepository.zip](https://github.com/user-attachments/files/16118326/InsuranceDataRepository.zip)

**Comment by oleg-shilo** [comment]
Have a look at this sample
![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/3c3faffe-1756-4017-b8bd-5234edec0009)

**Comment by abskulkarni** [comment]
> If you do not specify `Feature` then a default feature is created for automatically. It will be named `Complete`: ![image](https://private-user-images.githubusercontent.com/16729806/346909826-56ca58be-09c7-4638-b38a-f037c403a3d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1MjYzMzksIm5iZiI6MTcyMDUyNjAzOSwicGF0aCI6Ii8xNjcyOTgwNi8zNDY5MDk4MjYtNTZjYTU4YmUtMDljNy00NjM4LWIzOGEtZjAzN2M0MDNhM2Q3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA5VDExNTM1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcxZmZiZDY5YmVmYTgxOGJkMmY2ZTZlMzA3MzRhY2NlZGFlNGMyNGY2MGI3MmEyM2YyMGE0NTM2YmRhNzQ5YTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.M2R1zsCjD3OQyLojF7v_3BLp4qj_LgaNtZQhyNHP1-E)
> 
> The intended scenario is that you have a set of files to install and all of them belong to a feature (yours or a default one). Any other files are grouped in other features.
> 
> > If I install MSI, it installs empty "IDRDatabaseScripts" folder (which is a feature) even if it is not selected.
> 
> Sorry, debugging your production solution is not really a support option. If you share something you want me to run/debug it needs to be simplified and sanitized to the level of hello-world project. I hope you understand where I am coming from.

Hi @oleg-shilo 
thanks for your reply. I will try out "complete" feature related suggestion.
For other issue, I have not given full production setup. Source files are having simple empty test.txt files. I gave it as I am unable to locate this bug of extra specific empty folder. Hope you will help in resolving this.

**Comment by oleg-shilo** [comment]
Sorry @abskulkarni,
Without a simple hello-world style test case that can only provid with you generic guidance. Without readable test case I am blind.

Another suggestion, try set `AutoElements.SupportEmptyDirectories = CompilerSupportState.Disabled;`. It might be what you need.

**Comment by abskulkarni** [comment]
No problem @oleg-shilo 
I figured it out myself. I was taking specific extension files from base directory of feature  where those files with extensions were not present. (actual extension files are present in sub folders) So that feature was getting created as empty one. So its base folder was installed even if it has no files.

I don't know how successful I am in explaining my root cause but it is working now.

**Comment by oleg-shilo** [comment]
Actually, you are successful. :)
BTW, MSI/WiX out of the box does not install empty folders. WixSharp facilitates it via `SupportEmptyDirectories`. It was a feature implemented upon user request.
I am glad you figured it out.

**Comment by abskulkarni** [comment]
What is SupportEmptyDirectories ? Is it project property? 
So if my feature is an empty directory & if I set SupportEmptyDirectories to false, that feature will be ignored in feature list and not installed?

**Comment by oleg-shilo** [comment]
This is the statement that configures WixSharp engine to fake empty directories so MSI is happy. You set it this way:
```C#
AutoElements.SupportEmptyDirectories = CompilerSupportState.Disabled;
```

**Comment by oleg-shilo** [comment]
I have updated the documentation. And will schedule the partial merger of `AutoElements` and `Compiler.AutoGeneration`

**Comment by smaudet** [comment]
In addition, the examples project is wrong/misleading.

If you try to create a project with 2.1.2/2.17 at least, with only features, and a "default feature" which adds all those, the installer is un-usable (crashes).

**Comment by oleg-shilo** [comment]
> In addition, the examples project...

@smaudet, I assume you are referring to `<wixsharp>\Source\src\WixSharp.Samples\Wix# Samples\Features\setup.cs`. Is it right?
But this sample works OK. I just tested.
So can you elaborate on which sample crashes?

**Comment by smaudet** [comment]
@oleg-shilo No:

`Source/src/WixSharp.Samples/Wix# Samples/MergeModules/Consuming MergeModule/setup.cs`

A) there is what I presume to be a typo and featureB is included twice (instead of A and B) on lines 30,31.
B) when I was attempting to replicate this setup myself the msi would crash on startup. I had issues debugging and did not investigate further - I did not the "complete" feature didn't do anything useful and actually was the cause of the crash (somehow). It also shows up without being added explicitly.

**Comment by oleg-shilo** [comment]
Interesting.
The `featureB` is indeed included twice but on purpose. The idea was that two msm modules belong to the same feature. Though the `featureA` (with a text file deployment) line was commented out by accident. I have fixed it now:

```c#
var project =
        new Project("MyMergeModuleSetup",
            new MediaTemplate { CompressionLevel = CompressionLevel.none, EmbedCab = false },
            new Dir(@"%ProgramFiles%\My Company",
                new File(featureA, @"Files\MainFile.txt"), // was commented out
                new Merge(featureB, @"Files\MyMergeModule.msm"),
                new Merge(featureB, @"Files\MyMergeModule1.msm")),
            new EnvironmentVariable("foo", "bar"));
``` 

Having `featureA` commented out does not make the difference failure-wise, but you are right, without `featureA` the whole features split did not make sense. So it's fixed and the sample is already in the repo.

Now, about the crash.
It's strange, but the example is finishing OK for me:
![image](https://github.com/user-attachments/assets/faac451d-0758-4125-81e4-7bad911511a6)

MSM modules are sensitive to MSI Installer version. Is it possible that your target system somehow is incompatible? It's a wild guess. The log may give some idea. Can you collect the log file for that and share it?

And I tested the booth builds with and without `featureA` being commented out.

**Comment by smaudet** [comment]
@oleg-shilo I think I found the cause of my issue - the custom WPF UI template you include has several issues (lots of unsafe code), specifically the Nodes on Node are left null, which causes a problem when you attempt to load them...

I will create a new ticket in a minute, trying to fix my merge + node tree now...

**Comment by oleg-shilo** [comment]
yep, I have left the comment there for you. I think you might be missing the SHA-1: 96ff60253fcad0ecf7387c22684b4d90bf873c7e fix. But other problems may still be there. Will see...
